src/sinks/file.rs

```diff
@@ -0,0 +1,563 @@
+use crate::{
+    buffers::Acker,
+    event::{self, Event},
+    sinks::util::SinkExt,
+    template::Template,
+    topology::config::DataType,
+};
+
+use futures::{future, Async, AsyncSink, Sink, StartSend};
+use serde::{Deserialize, Serialize};
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+use std::time::Instant;
+
+#[derive(Deserialize, Serialize, Debug, Eq, PartialEq, Clone)]
+#[serde(rename_all = "snake_case")]
+pub enum Encoding {
+    Text,
+    Json,
+}
+
+#[derive(Deserialize, Serialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct FileSinkConfig {
+    pub path: String,
+    pub close_timeout_secs: Option<u64>,
+    pub encoding: Option<Encoding>,
+}
+
+impl FileSinkConfig {
+    pub fn new(path: String) -> Self {
+        Self {
+            path,
+            close_timeout_secs: None,
+            encoding: None,
+        }
+    }
+}
+
+#[typetag::serde(name = "file")]
+impl crate::topology::config::SinkConfig for FileSinkConfig {
+    fn build(&self, acker: Acker) -> Result<(super::RouterSink, super::Healthcheck), String> {
+        let sink = PartitionedFileSink::new(
+            Template::from(&self.path[..]),
+            self.close_timeout_secs,
+            self.encoding.clone(),
+        )
+        .stream_ack(acker);
+
+        Ok((Box::new(sink), Box::new(future::ok(()))))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+}
+
+pub type EmbeddedFileSink = Box<dyn Sink<SinkItem = Event, SinkError = ()> + 'static + Send>;
+
+pub struct PartitionedFileSink {
+    path: Template,
+    encoding: Option<Encoding>,
+    close_timeout_secs: Option<u64>,
+
+    partitions: HashMap<Arc<Path>, EmbeddedFileSink>,
+    last_accessed: HashMap<Arc<Path>, Instant>,
+    closing: Vec<EmbeddedFileSink>,
+}
+
+impl PartitionedFileSink {
+    pub fn new(
+        path: Template,
+        close_timeout_secs: Option<u64>,
+        encoding: Option<Encoding>,
+    ) -> Self {
+        PartitionedFileSink {
+            path,
+            encoding,
+            close_timeout_secs,
+            partitions: HashMap::new(),
+            last_accessed: HashMap::new(),
+            closing: Vec::new(),
+        }
+    }
+
+    fn collect_old_files(&mut self) {
```

Comment:
this function calls out to other poll_*, I would like to see this renamed to `poll_collect_files` to indicate that this does task context based work. This should make it easier to approach this sink later on and understand what these functions do.

Reply:
fixed